### DISCUSSION TITLE
AI needs a delay to try to re-open the door.

### DISCUSSION THREAD

**Discussion by MaaiG** [FakeFishGames/Barotrauma#7424] [Ideas and suggestions]
In my case, the bot can’t use my airlock chamber .
After pressing the button you have to wait for the airlock chamber to fill water - then the door will open.
The bot presses the button, but the door does not open immediately. The bot presses the button again and again and the airlock is shuts down.
Then the bot constantly tries to press the button (as if in a panic) and gets into the loop. The sequence does not work because the bot constantly switches the button and interrupts it.

I think it might be nice if AI will have a delay between trying to open the door! Or number of attempts will be limited and odd! ))

**Comment by itchyOwl** [comment]
There's now 3 second delay after successfully pressing a button before trying again (Regaliss11/Barotrauma-development@e8a33a1, included in the next unstable). But your case is very specific. How should the bots know that they should wait before the chamber fills? Maybe you could try using a delay or some logic component to prevent the issue?

**Reply by MaaiG** [reply]
> How should the bots know that they should wait before the chamber fills?

I dont know))
In fact, why the bots press the button again? Afraid to miss? Or for what?

>Maybe you could try using a delay or some logic component to prevent the issue?

Need to think "how to set the delay for the second signal", but then the cycle cannot be interrupted if the need arises.

**Comment by itchyOwl** [comment]
The bots only attempt to open doors that they should have access to. And in general they should press the button only once, but in practice there's multiple things that can make the first attempt fail. For example someone else closes the door just before them, or an automated system does that. So they need to be able to cope with that. It's basically the same you would probably do if nobody came to the door after you ringed a bell, supposing that you know there's a person inside that should open it: ring again. And even in a perfect world they would have to reset after having opened the door once, or they wouldn't be able to open that door ever during the session again. So that reset now happens after 3 seconds. Previously they did just try to press all the buttons (they thought they need to) every 0.25 seconds.

**Reply by MaaiG** [reply]
> Previously they did just try to press all the buttons (they thought they need to) every 0.25 seconds.

Thanks! Three seconds that is a lot better! That is the time for fill a small airlock. In principle, you can accelerate the time to fill... Anyway, it’s a lot better than "hydro strike" when you trying to get out.

>So they need to be able to cope with that. It's basically the same you would probably do if nobody came to the door after you ringed a bell, supposing that you know there's a person inside that should open it: ring again.

Yes, but to be honest, it’s a rigid framework. It’s robots. They don’t know anything. They can only wait, and then they can leave or they can look for another way. Maybe they will try to use a crowbar. )) They shouldn’t just bang on the door all the time. In real life, normal people don’t do that. If your attempts are fruitless, you stop and make a different choice.
- So I thought about making to limit the number of attempts for one action.

>And in general they should press the button only once, but in practice there's multiple things that can make the first attempt fail.

- So I thought about making the number of attempts odd.

There are toys that are constantly crashing into the same obstacle, and there are toys that overcome the obstacle. The second has a rigid framework too, but more choices.
If we don’t think about ourselves as robots, then maybe we’re very naive. ))

PS 

>How should the bots know that they should wait before the chamber fills?

Perhaps the `buttons` could have "settings (flag or etc)" that the bots could read and understand how to work with them.
Or as a simple "router". Like a traffic light with yellow light (the button is in the process of completing the task). ;)

**Comment by itchyOwl** [comment]
You are right that they could be smarter. But it's a question of whether we should use the resources in improving their behavior for each corner case or not. A game AI is a bit easier than real world AI (e.g. cleaning robots), because the bots have much better access to the data of the world. There's much less uncertainties. So they don't have to be as smart. Here it's a custom system that they don't quite understand.

So instead of limiting the attempts of doing something they should normally be able to do (which might cause issues elsewhere), I'd consider adding that automated logic in the doors. For example, the innermost door of Kastrull's airlock does not open when the airlock is filled with water. The bots try to press the button, but it does nothing when there's water. And it works:
![image](https://user-images.githubusercontent.com/5370738/144271819-f9fffec7-955e-474c-a849-6d0975ad33eb.png)

![image](https://user-images.githubusercontent.com/5370738/144272329-6fcda11e-0365-47a3-b84e-622a7813cfd4.png)

**Reply by MaaiG** [reply]
> But it's a question of whether we should use the resources in improving their behavior for each corner case or not. 
I think an integrated approach is always needed. An individual approach is needed only in special cases. Where there’s no other way.

> A game AI is a bit easier than real world AI (e.g. cleaning robots), because the bots have much better access to the data of the world. There's much less uncertainties. So they don't have to be as smart.

Yea! Otherwise, they may suffer from the impossibility of self-realization. ))

> So instead of limiting the attempts of doing something they should normally be able to do (which might cause issues elsewhere), I'd consider adding that automated logic in the doors.

Yea, thanks. No problem. I moved quickly to the stage of creating my own submarines. I like the sub-editor sandbox.
If the bots can’t get their level closer to mine, I can always get down to their level. )) But it would be interesting.
Three seconds instead of 0.25 - is what is needed! Thanks! Now there’s a place to start.

**Reply by MaaiG** [reply]
Nooo... let the bots be smarter! :)

**Comment by MaaiG** [comment]
Related [topic](https://github.com/Regalis11/Barotrauma/discussions/8148)!
